[SPARK-14647][SQL] Group SQLContext/HiveContext state into SharedState

andrewor14 opened the pull request:
## What changes were proposed in this pull request?

This patch adds a `PersistentState` that groups state shared across multiple `SQLContext`s. This is analogous to the `SessionState` added in SPARK-13526 that groups session-specific state. This cleanup makes the constructors of the contexts simpler and ultimately allows us to remove `HiveContext` in the near future.
## How was this patch tested?

Just refactoring.

Commit:
Group SQLContext constructor args into PersistentState

Commit:
Move the HiveClient's into SessionState

SparkQA commented:
**[Test build #55881 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55881/consoleFull)** for PR 12405 at commit [`e8c78c3`](https://github.com/apache/spark/commit/e8c78c308b3e6cbd23843b9acd4f79e594c4835e).

Commit:
Fix style

SparkQA commented:
**[Test build #55881 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55881/consoleFull)** for PR 12405 at commit [`e8c78c3`](https://github.com/apache/spark/commit/e8c78c308b3e6cbd23843b9acd4f79e594c4835e).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55881/
Test FAILed.

SparkQA commented:
**[Test build #55882 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55882/consoleFull)** for PR 12405 at commit [`a968aaf`](https://github.com/apache/spark/commit/a968aaf9331df307b9750c12ea69706f121e1a2c).

Commit:
Always use HiveClient (not Impl)

Previously executionHive was a HiveClientImpl but metadataHive
was a HiveClient. This is inconsistent and somewhat arbitrary.
It's better practice to always use the abstract thing.

rxin commented:
We need to find a different name -- the states are not necessarily persistent (e.g. cache manager).

SparkQA commented:
**[Test build #55883 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55883/consoleFull)** for PR 12405 at commit [`e0a3c94`](https://github.com/apache/spark/commit/e0a3c94709fd7ba9f455622a6d72aa9fbf7ae2fd).

SparkQA commented:
**[Test build #55883 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55883/consoleFull)** for PR 12405 at commit [`e0a3c94`](https://github.com/apache/spark/commit/e0a3c94709fd7ba9f455622a6d72aa9fbf7ae2fd).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55883/
Test FAILed.

SparkQA commented:
**[Test build #55882 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55882/consoleFull)** for PR 12405 at commit [`a968aaf`](https://github.com/apache/spark/commit/a968aaf9331df307b9750c12ea69706f121e1a2c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55882/
Test FAILed.

yhuai commented:
```
2016-04-14 18:24:27.954 - stderr> Exception in thread "main" java.lang.NullPointerException
2016-04-14 18:24:27.954 - stderr> 	at org.apache.spark.sql.SQLContext.sparkContext(SQLContext.scala:100)
2016-04-14 18:24:27.954 - stderr> 	at org.apache.spark.sql.SQLContext.<init>(SQLContext.scala:81)
2016-04-14 18:24:27.954 - stderr> 	at org.apache.spark.sql.hive.HiveContext.<init>(HiveContext.scala:83)
2016-04-14 18:24:27.955 - stderr> 	at org.apache.spark.sql.hive.HiveContext.<init>(HiveContext.scala:88)
2016-04-14 18:24:27.955 - stderr> 	at org.apache.spark.sql.hive.thriftserver.SparkSQLEnv$.init(SparkSQLEnv.scala:59)
2016-04-14 18:24:27.955 - stderr> 	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.<init>(SparkSQLCLIDriver.scala:281)
2016-04-14 18:24:27.955 - stderr> 	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver$.main(SparkSQLCLIDriver.scala:138)
2016-04-14 18:24:27.961 - stderr> 	at org.apache.spark.sql.hive.thriftserver.SparkSQLCLIDriver.main(SparkSQLCLIDriver.scala)
2016-04-14 18:24:27.961 - stderr> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
2016-04-14 18:24:27.961 - stderr> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
2016-04-14 18:24:27.962 - stderr> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
2016-04-14 18:24:27.962 - stderr> 	at java.lang.reflect.Method.invoke(Method.java:497)
2016-04-14 18:24:27.962 - stderr> 	at org.apache.spark.deploy.SparkSubmit$.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:726)
2016-04-14 18:24:27.962 - stderr> 	at org.apache.spark.deploy.SparkSubmit$.doRunMain$1(SparkSubmit.scala:183)
2016-04-14 18:24:27.962 - stderr> 	at org.apache.spark.deploy.SparkSubmit$.submit(SparkSubmit.scala:208)
2016-04-14 18:24:27.962 - stderr> 	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:122)
2016-04-14 18:24:27.962 - stderr> 	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
```

andrewor14 commented:
Hm, `SharedState`?

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveContext.scala:
```diff
@@ -114,25 +94,25 @@ class HiveContext private[hive](
   logDebug("create HiveContext")
 
+  @transient
+  protected[sql] override val persistentState: PersistentState = hivePersistentState
```
This line caused the NPE. We can just remove it, right?

yhuai commented:
`SharedState` sounds better.

andrewor14 commented:
`GlobalState`

rxin commented:
either works --

Commit:
Fix tests?

Commit:
PersistentState -> SharedState

Commit:
Merge branch 'master' of github.com:apache/spark into persistent-state

Commit:
Fix style

SparkQA commented:
**[Test build #55903 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55903/consoleFull)** for PR 12405 at commit [`5539c1b`](https://github.com/apache/spark/commit/5539c1bc36f2e59b14f4836dd5f5b4423dc4ad0a).

SparkQA commented:
**[Test build #55903 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55903/consoleFull)** for PR 12405 at commit [`5539c1b`](https://github.com/apache/spark/commit/5539c1bc36f2e59b14f4836dd5f5b4423dc4ad0a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55903/
Test FAILed.

Commit:
Revert "Always use HiveClient (not Impl)"

This reverts commit e0a3c94709fd7ba9f455622a6d72aa9fbf7ae2fd.

SparkQA commented:
**[Test build #55942 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55942/consoleFull)** for PR 12405 at commit [`c540a9c`](https://github.com/apache/spark/commit/c540a9c3a94ed6448bbbe854fe0aaafd1d37fae5).

SparkQA commented:
**[Test build #55942 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55942/consoleFull)** for PR 12405 at commit [`c540a9c`](https://github.com/apache/spark/commit/c540a9c3a94ed6448bbbe854fe0aaafd1d37fae5).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55942/
Test FAILed.

Commit:
Merge branch 'master' of github.com:apache/spark into persistent-state

Commit:
Fix tests?

SparkQA commented:
**[Test build #55964 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55964/consoleFull)** for PR 12405 at commit [`9c73146`](https://github.com/apache/spark/commit/9c73146c2f0a278759401682a18f1ad5e26f81a0).

SparkQA commented:
**[Test build #55964 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55964/consoleFull)** for PR 12405 at commit [`9c73146`](https://github.com/apache/spark/commit/9c73146c2f0a278759401682a18f1ad5e26f81a0).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55964/
Test FAILed.

andrewor14 commented:
retest this please

SparkQA commented:
**[Test build #55966 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55966/consoleFull)** for PR 12405 at commit [`9c73146`](https://github.com/apache/spark/commit/9c73146c2f0a278759401682a18f1ad5e26f81a0).

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/test/TestHive.scala:
```diff
@@ -549,6 +505,22 @@ private[hive] class TestHiveFunctionRegistry extends SimpleFunctionRegistry {
+    metastoreTemporaryConf: Map[String, String])
+  extends HiveSharedState(sc) {
+
+  override val metadataHive: HiveClient = {
```
I figured out the issue with the previous test failure. Although we are trying to override it at here, but inside HiveSharedState, we still create a metadataHive which uses the derby `metastore_db`.

Commit:
Fix tests?

SparkQA commented:
**[Test build #55973 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55973/consoleFull)** for PR 12405 at commit [`2a84b8c`](https://github.com/apache/spark/commit/2a84b8c6bc5e62bd4eaee03c7623f3095c3b9698).

SparkQA commented:
**[Test build #55966 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55966/consoleFull)** for PR 12405 at commit [`9c73146`](https://github.com/apache/spark/commit/9c73146c2f0a278759401682a18f1ad5e26f81a0).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55966/
Test FAILed.

andrewor14 commented:
retest this please

SparkQA commented:
**[Test build #55982 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55982/consoleFull)** for PR 12405 at commit [`2a84b8c`](https://github.com/apache/spark/commit/2a84b8c6bc5e62bd4eaee03c7623f3095c3b9698).

yhuai commented:
looks like a legitimate failure?

SparkQA commented:
**[Test build #55973 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55973/consoleFull)** for PR 12405 at commit [`2a84b8c`](https://github.com/apache/spark/commit/2a84b8c6bc5e62bd4eaee03c7623f3095c3b9698).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55973/
Test PASSed.

SparkQA commented:
**[Test build #55982 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/55982/consoleFull)** for PR 12405 at commit [`2a84b8c`](https://github.com/apache/spark/commit/2a84b8c6bc5e62bd4eaee03c7623f3095c3b9698).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/55982/
Test PASSed.

yhuai commented on sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSharedState.scala:
```diff
@@ -0,0 +1,52 @@
+  /**
+   * A Hive client used to interact with the metastore.
+   */
+  lazy val metadataHive: HiveClient = {
```
Let's explain why we need to make it a lazy val at here.
